src/wx/widgets/sdljoy.cpp

```diff
@@ -38,420 +36,455 @@ static int16_t axisval(int16_t x)
-                    auto val      = axisval(SDL_JoystickGetAxis(joy.second.dev, axis));
-                    auto prev_val = joy.second.axis[axis];
+        case SDL_JOYDEVICEADDED:
+        {
+            int index = e.jdevice.which;
 
-                    if (val != prev_val) {
-                        CreateAndSendEvent(joy.first, WXSDLJOY_AXIS, axis, val, prev_val);
+            // Do not process joystick events for game controllers.
+            if (SDL_IsGameController(index))
+                return;
 
-                        joy.second.axis[axis] = val;
+            // Only add requested indexes.
+            if (add_all_joysticks_ || contains(requested_indexes_, index))
+                ConnectController(index);
 
-                        wxLogDebug("POLLED SDL_JOYAXISMOTION: joy:%d axis:%d val:%d prev_val:%d", joy.first, axis, val, prev_val);
-                    }
-                }
-            }
+            got_event = true;
+            break;
         }
-    }
-}
 
-void wxSDLJoy::ConnectController(uint8_t joy)
-{
-    SDL_Joystick* js_dev = nullptr;
-    bool is_gc;
-
-    if ((is_gc = SDL_IsGameController(joy))) {
-        auto dev = SDL_GameControllerOpen(joy);
-
-        if (dev) {
-            joystate[joy].dev = dev;
+        case SDL_JOYDEVICEREMOVED:
+        {
+            SDL_JoystickID joy_id = e.jdevice.which;
+            joystick_states_.erase(joy_id);
+            got_event = true;
+            break;
+        }
 
-            js_dev = SDL_GameControllerGetJoystick(dev);
+        default:
+            // Ignore all other events.
+            break;
         }
     }
-    else {
-        if ((js_dev = SDL_JoystickOpen(joy)))
-            joystate[joy].dev = js_dev;
-    }
 
-    if (js_dev) {
-        auto instance = SDL_JoystickInstanceID(js_dev);
+    bool do_poll = false;
+    wxLongLong now = wxGetUTCTimeMillis();
 
-        instance_map[instance] = &(joystate[joy]);
-
-        joystate[joy].instance = instance;
+    if (got_event) {
+        last_poll_ = now;
+    } else if (now - last_poll_ > POLL_TIME_MS) {
+        do_poll = true;
+        last_poll_ = now;
     }
 
-    joystate[joy].index = joy;
-    joystate[joy].is_gc = is_gc;
-}
-
-void wxSDLJoy::RemapControllers()
-{
-    for (auto&& joy : joystate) {
-        auto& state = joy.second;
-
-        DisconnectController(state);
-        ConnectController(joy.first);
+    if (do_poll) {
+        for (auto&& joy_state : joystick_states_) {
+            joy_state.second->Poll();
+        }
     }
 }
 
-void wxSDLJoy::DisconnectController(wxSDLJoyState& state)
-{
-    if (auto& dev = state.dev) {
-        if (state.is_gc) {
-            SDL_GameControllerClose(dev);
-        }
-        else {
-            SDL_JoystickClose(dev);
-        }
-
-        dev = nullptr;
+void wxSDLJoy::ConnectController(int index) {
+    std::unique_ptr<wxSDLJoyState> joy_state(new wxSDLJoyState(index));
+    if (joy_state->IsValid()) {
+        joystick_states_.emplace(
```

Comment:
This should be done by first erasing the index and then inserting the new state.

But since SDL likely handle multiple open calls just fine, just replace the `emplace()` call with an `erase()` followed by `insert()`.

Reply:
Actually no, that won't work either, if we do that the device will be closed and the new state will no longer be valid.

You'll need to keep a map of indexes to states and erase the connecting index and then insert it.

Reply:
Hi @rkitover, also consider this point made by SDL:

```
2. Joystick and controller are ref counted. You need to call close for each call to open to actually free their memory.
```

We should avoid calling open/close multiple times.  I think this is used for a publish/subscriber type of behavior.  For example, let's say multiple modules want to keep track of the Joystick independently.  It should only be freed when they are all finished.  I don't think we have this need, so we should only call open once.

Reply:
I was thinking about my comment more, and I think it this strategy would work if we also use player_index instead of index, like you are now.  

So, my vote is for:

Not closing/reopening a controller on every state change
Using player_index, as it isn't as volatile as index

Reply:
I spent some time toying around with SDL and indexes and everything. Unfortunately, there is no easy fix for this. I dhave done some changes to deal with this:
* The `wxSDLJoy` public API should only deal with *player indexes*. The fact that we also use *SDL indexes* internally is an implementation details that is of no use to the caller.
* As a result, `wxSDLJoy` should take care of converting between player indexes and SDL indexes.
* The code should be clear on what it is actually using. I have renamed variables and parameters to that end. Player indexes are unsigned, SDL indexes are signed because that is what the SDL API uses. Unfortunately, I have not figured out a good way to map between the 2 other than adding or substracting 1, which is ugly and probably error-prone down the line. However, as a result, I no longer care about the SDL index after joystick initialization, which is a slight improvement.
* We basically have 2 modes for connecting joysticks: During configuration, where we poll all joysticks and otherwise, where we only poll a select number of joysticks. I have changed the public API to reflect this. We now have a `PollJoysticks()` API to poll a select number of player indexes. and `PollAllJoysticks()` where we remap on every connect/disconnect. The reason for this is that we do not want the active joystick to disconnect during gameplay because another, unrelated joystick has disconnected. For instance, if you are playing with joystick player index 2 (SDL index 1), you do not want to be disconnected because joystick player index 1 (SDL index 0) has disconnected.
* SDK is *weird* with joystick ID handling. Connect 2 controllers (SDL indexes 0 and 1), disconnect SDL index 0, you are left with one controller, that ends up having SDL index 0. Reconnect the other controller, both controllers are back to their original SDL index position, meaning the previous controller is back to SDL index 1.

Please let me know what you think. I'm not very happy with my solution but I can't figure out a better way to do this without losing the joystick connection during gameplay in some cases - like multiple VBA-M instances running at the same time.

Reply:
I think I am on the same page that you are.  I tried a few things, similar to you, and found that there is no *perfect* solution here.  Of course we are at the mercy of the SDL library, so it looks like the best we can do is to agree that some solution is our best possible effort.  Let me just respond to your points with some of my thoughts:

>The wxSDLJoy public API should only deal with player indexes. The fact that we also use SDL indexes internally is an implementation details that is of no use to the caller.

This is correct.  I would actually argue we should not use the SDL joystick index at all past the initially connection.  I don't think storing them is valuable to us.

For example, in my implementation I was thinking the player_index should be managed as follows:
We have a maximum of N players (say, 4 players).  We have a simple boolean array of length 4:

[false, false, false, false]

This indicates to us if we have any players active on that player_index.  Now, we connect the first controller, it has SDL index 0, instance ID 1.  We assign it to player_index 0 using SDL_GameControllerSetPlayerIndex.

[true, false, false, false]

Now, we accept another controller connection.  Our second player is SDL index 1, instance ID 2.  We look at our map, and determine that player_index 1 is available.  We set this controller to player_index 1 using SDL_GameControllerSetPlayerIndex.

[true, true, false, false]

Now, say player 1 disconnects.  Sure, the SDL indexes will be sliding around.  Player 2's SDL index will slide from index 1 to index 0, like you described.  However, it will still have player_index 1.  This should be the source of truth, along with its instance ID.

After player 1 disconnects, we clear that hole in our table:

[false, true, false, false]

Now, whatever controller connects next, will get player_index 0, and fill the hole we created.

[true, true, false, false]

So, my idea is to ditch SDL index immediately, set a self determined player index via SDL_GameControllerSetPlayerIndex and SDL_GameControllerGetPlayerIndex to avoid sliding controllers.  This is basically a very rudimentary controller port system, like on a console.  The only downside to this is that we have no concept of tracking controllers through disconnects, so technically if both player 3 and player 1 disconnect, they could swap depending on the order they reconnect in.

I just read now that you recently introduced player_index instead of index, which is basically what I'm describing here.  However, I see it is simply set to SDL index + 1.  Can you explain what your idea is now?

> As a result, wxSDLJoy should take care of converting between player indexes and SDL indexes.

Right, this is basically the system I described above.  It's simply one idea, there are other ways of doing this like you've probably thought of.

> The code should be clear on what it is actually using. I have renamed variables and parameters to that end. Player indexes are unsigned, SDL indexes are signed because that is what the SDL API uses. Unfortunately, I have not figured out a good way to map between the 2 other than adding or substracting 1, which is ugly and probably error-prone down the line. However, as a result, I no longer care about the SDL index after joystick initialization, which is a slight improvement.

Great, I am a fan of all of this.  I think not caring about it is the right way to go.  And the more accurate naming will help us kill the bugs more easily, because we should never confused the context in which they are used.

>  We basically have 2 modes for connecting joysticks: During configuration, where we poll all joysticks and otherwise, where we only poll a select number of joysticks. I have changed the public API to reflect this. We now have a PollJoysticks() API to poll a select number of player indexes. and PollAllJoysticks() where we remap on every connect/disconnect. The reason for this is that we do not want the active joystick to disconnect during gameplay because another, unrelated joystick has disconnected. For instance, if you are playing with joystick player index 2 (SDL index 1), you do not want to be disconnected because joystick player index 1 (SDL index 0) has disconnected.

I was a little confused by this, how could this happen?  Would we just receive an event that index 2 disconnected, and then close that connection?  How could we accidentally disconnect player index 2 in this case?

> SDK is weird with joystick ID handling. Connect 2 controllers (SDL indexes 0 and 1), disconnect SDL index 0, you are left with one controller, that ends up having SDL index 0. Reconnect the other controller, both controllers are back to their original SDL index position, meaning the previous controller is back to SDL index 1.

Yes, I'm assuming you're referring to SDL index and not player index, this is also kind of why I had my idea to track them internally using some data structure (I've been calling this behavior "sliding joysticks" in this post).

>  Please let me know what you think. I'm not very happy with my solution but I can't figure out a better way to do this without losing the joystick connection during gameplay in some cases - like multiple VBA-M instances running at the same time.

I will also think on this a bit.  I have not fully reviewed your code yet, but I suggest we just think about it for a day or two and just decide what is the best course of action.  It doesn't need to be perfect, but since you already did a great job, let's make it as great as we can.